### PR TITLE
AD20-16: Support to cancel docker execution

### PR DESCRIPTION
**Pull request by hernaldourbina**
Cancel/stop feature. There was an endpoint for handling cancel feature, cancel already works for tarball mode but it is not implemented yet for docker.

### CODE REVIEW

File: engine/src/main/java/com/odysseusinc/arachne/executionengine/execution/r/DockerOverseer.java [hunk 1]
```diff
@@ -58,17 +59,29 @@ public DockerOverseer(
...
-                new ExecutionOutcome(Stage.ABORT, "Abort is not yet supported for docker analysis", null)
-        );
+        return CompletableFuture.supplyAsync(() -> {
+            if (result.isDone()) {
+                return result.join();
+            } else {
+                try {
+                    client.stopContainerCmd(containerId).exec();
+                    client.removeContainerCmd(containerId).withRemoveVolumes(true).exec();
+                    return new ExecutionOutcome(Stage.ABORT, "Docker container aborted successfully", null);
```
Comment (dmitrys-odysseus) [inline engine/src/main/java/com/odysseusinc/arachne/executionengine/execution/r/DockerOverseer.java:78]: Defect: empty stdout in 3rd argument will overwrite everything in datanode. ExecutionOutcome must return FULL stdout.
Defect: this is not a correct termination state. First argument is stage and second argument is error that happend on that staget preventing the process from progressing. So, stage="ABORT" / error="something" signals that aborting has failed. A correct  outcome here is stage="ABORTED" / error=null. The message about successful abort should be placed in the stdout.

File: engine/src/main/java/com/odysseusinc/arachne/executionengine/execution/r/DockerOverseer.java [hunk 2]
```diff
@@ -58,17 +59,29 @@ public DockerOverseer(
...
+                } catch (Exception e) {
+                    log.error("Error stopping the Docker container: {}", e.getMessage());
+                    throw new RuntimeException("Error aborting Docker container: " + e.getMessage());
```
Comment (dmitrys-odysseus) [inline engine/src/main/java/com/odysseusinc/arachne/executionengine/execution/r/DockerOverseer.java:81]: Defect: Exception should not be thrown here.
Now, this is the case when outcome with stage="ABORT" / error="something" should be produced

File: engine/src/main/java/com/odysseusinc/arachne/executionengine/execution/r/DockerOverseer.java [hunk 3]
```diff
@@ -58,17 +59,29 @@ public DockerOverseer(
...
-        );
+        return CompletableFuture.supplyAsync(() -> {
+            if (result.isDone()) {
+                return result.join();
+            } else {
+                try {
+                    client.stopContainerCmd(containerId).exec();
+                    client.removeContainerCmd(containerId).withRemoveVolumes(true).exec();
+                    return new ExecutionOutcome(Stage.ABORT, "Docker container aborted successfully", null);
+                } catch (Exception e) {
```
Comment (dmitrys-odysseus) [inline engine/src/main/java/com/odysseusinc/arachne/executionengine/execution/r/DockerOverseer.java:79]: Generic catch? if we already have `exceptionally(...)` below for a generic fallback, I don't quite get what arewe anticipating here.

**Review by dmitrys-odysseus — CHANGES_REQUESTED** [review]

File: engine/src/main/java/com/odysseusinc/arachne/executionengine/execution/r/DockerOverseer.java [hunk 4]
```diff
@@ -58,17 +59,29 @@ public DockerOverseer(
...
-        return result.isDone() ? result : CompletableFuture.completedFuture(
-                new ExecutionOutcome(Stage.ABORT, "Abort is not yet supported for docker analysis", null)
-        );
+        return CompletableFuture.supplyAsync(() -> {
+            if (result.isDone()) {
+                return result.join();
+            } else {
+                try {
+                    client.stopContainerCmd(containerId).exec();
+                    client.removeContainerCmd(containerId).withRemoveVolumes(true).exec();
```
[the remark this answers is not in the document]
Comment (hernaldourbina) [reply engine/src/main/java/com/odysseusinc/arachne/executionengine/execution/r/DockerOverseer.java]: Can we click stop once the docker has been completed?

File: engine/src/main/java/com/odysseusinc/arachne/executionengine/execution/r/DockerOverseer.java [hunk 5]
```diff
@@ -58,17 +59,27 @@ public DockerOverseer(
...
-        return result.isDone() ? result : CompletableFuture.completedFuture(
-                new ExecutionOutcome(Stage.ABORT, "Abort is not yet supported for docker analysis", null)
-        );
+        return CompletableFuture.supplyAsync(() -> {
+            if (result.isDone()) {
+                return result.join();
+            } else {
+                client.stopContainerCmd(containerId).exec();
+                client.removeContainerCmd(containerId).withRemoveVolumes(true).exec();
+                return new ExecutionOutcome(Stage.ABORT, null, "Docker container aborted successfully");
```
Comment (dmitrys-odysseus) [inline engine/src/main/java/com/odysseusinc/arachne/executionengine/execution/r/DockerOverseer.java:77]: The stdout in ExecutionOutcome is be FULL stdout. Not delta.

File: engine/src/main/java/com/odysseusinc/arachne/executionengine/execution/r/DockerOverseer.java [hunk 6]
```diff
@@ -58,17 +59,27 @@ public DockerOverseer(
...
+            }
+        }).exceptionally(e -> {
+            log.error("Error stopping the Docker container: {}", e.getMessage());
+            return new ExecutionOutcome(Stage.ABORTED, "Error aborting Docker container: " + e.getMessage(), null);
```
Comment (dmitrys-odysseus) [inline engine/src/main/java/com/odysseusinc/arachne/executionengine/execution/r/DockerOverseer.java:81]: stdout in ExecutionOutcome is missing

File: engine/src/main/java/com/odysseusinc/arachne/executionengine/execution/r/DockerOverseer.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (dmitrys-odysseus) [reply engine/src/main/java/com/odysseusinc/arachne/executionengine/execution/r/DockerOverseer.java]: That doesn't matter. 
Occam's Razor: If this line has a meaning, explain or demonstrate it. Otherwise, it should not be added.

File: engine/src/main/java/com/odysseusinc/arachne/executionengine/execution/r/DockerOverseer.java [hunk 7]
```diff
@@ -36,6 +36,8 @@ public class DockerOverseer extends AbstractOverseer {
      */
     private volatile int pos;
     private final DockerClient client;
+    private String containerId;
+    private static final StringBuilder executionLogs = new StringBuilder();
```
Comment (dmitrys-odysseus) [inline engine/src/main/java/com/odysseusinc/arachne/executionengine/execution/r/DockerOverseer.java:40]: There is already a field `stdout` that contains all the execution logs. Why do you need to create another one?

File: engine/src/main/java/com/odysseusinc/arachne/executionengine/execution/r/DockerOverseer.java [hunk 8]
```diff
@@ -58,17 +60,27 @@ public DockerOverseer(
         }));
         pos = stdout.length();
         this.client = client;
-        init.thenAccept(containerId ->
-                executor.scheduleWithFixedDelay(this::writeLogs, updateInterval, updateInterval, TimeUnit.MILLISECONDS)
-        );
+        init.thenAccept(containerId -> {
+            this.containerId = containerId;
```
Comment (dmitrys-odysseus) [inline engine/src/main/java/com/odysseusinc/arachne/executionengine/execution/r/DockerOverseer.java:64]: containerId is _already stored_ in this class in `init` field - as soon as it is there. Storing it again is excessive. Where you need it, initi field shoulf be used.

File: engine/src/main/java/com/odysseusinc/arachne/executionengine/execution/r/DockerOverseer.java [hunk 9]
```diff
@@ -58,17 +60,27 @@ public DockerOverseer(
...
     public CompletableFuture<ExecutionOutcome> abort() {
-        return result.isDone() ? result : CompletableFuture.completedFuture(
-                new ExecutionOutcome(Stage.ABORT, "Abort is not yet supported for docker analysis", null)
-        );
+        String logs = getExecutionLogs();
+        return CompletableFuture.supplyAsync(() -> {
+            if (result.isDone()) {
+                return result.join();
+            } else {
+                client.stopContainerCmd(containerId).exec();
```
Comment (dmitrys-odysseus) [inline engine/src/main/java/com/odysseusinc/arachne/executionengine/execution/r/DockerOverseer.java:77]: Major: If `init` has not yet completed (e.g. the container is still being downloaded), this will fail with NPE and then, container will wind up and proceed to completion, effectively ignoring the abort command.

File: engine/src/main/java/com/odysseusinc/arachne/executionengine/execution/r/DockerOverseer.java [hunk 10]
```diff
@@ -58,17 +60,27 @@ public DockerOverseer(
...
+                return new ExecutionOutcome(Stage.ABORT, null, "Docker container aborted successfully:\n" + logs);
+            }
+        }).exceptionally(e -> {
+            log.error("Error stopping the Docker container: {}", e.getMessage());
+            return new ExecutionOutcome(Stage.ABORTED, "Error aborting Docker container: " + e.getMessage(), "Error aborting Docker container: " + e.getMessage() + "\n" + logs);
```
Comment (dmitrys-odysseus) [inline engine/src/main/java/com/odysseusinc/arachne/executionengine/execution/r/DockerOverseer.java:82]: Here, stage should be ABORT, not ABORTED

File: engine/src/main/java/com/odysseusinc/arachne/executionengine/execution/r/DockerOverseer.java (hunk 9 above)
Comment (dmitrys-odysseus) [inline engine/src/main/java/com/odysseusinc/arachne/executionengine/execution/r/DockerOverseer.java:77]: According to the documentation, there are dedicated conditions for stopContainerCmd:
```
NotFoundException - No such container
NotModifiedException - Container already stopped
```
There error conditions (well, at least one of them) should be handled explicitly as idempotent success.
Yes, this is tricky as completion of a container can race with completion of the `result` future and the code has to account for that.

File: engine/src/main/java/com/odysseusinc/arachne/executionengine/execution/r/DockerOverseer.java [hunk 11]
```diff
@@ -58,17 +60,27 @@ public DockerOverseer(
...
-        return result.isDone() ? result : CompletableFuture.completedFuture(
-                new ExecutionOutcome(Stage.ABORT, "Abort is not yet supported for docker analysis", null)
-        );
+        String logs = getExecutionLogs();
+        return CompletableFuture.supplyAsync(() -> {
+            if (result.isDone()) {
+                return result.join();
+            } else {
+                client.stopContainerCmd(containerId).exec();
+                return new ExecutionOutcome(Stage.ABORT, null, "Docker container aborted successfully:\n" + logs);
```
Comment (dmitrys-odysseus) [inline engine/src/main/java/com/odysseusinc/arachne/executionengine/execution/r/DockerOverseer.java:78]: ABORT is in-progress stated. Successful completion on this path is ABROTED

**Review by dmitrys-odysseus — CHANGES_REQUESTED** [review]

File: engine/src/main/java/com/odysseusinc/arachne/executionengine/execution/r/DockerOverseer.java (hunk 7 above)
Comment (hernaldourbina) [reply engine/src/main/java/com/odysseusinc/arachne/executionengine/execution/r/DockerOverseer.java:40]: You are right!.

File: engine/src/main/java/com/odysseusinc/arachne/executionengine/execution/r/DockerOverseer.java (hunk 9 above)
Comment (hernaldourbina) [reply engine/src/main/java/com/odysseusinc/arachne/executionengine/execution/r/DockerOverseer.java:77]: Thanks for catching this up. To handle this case I believe we should wait for its completion before proceeding with the abort logic. We will use the `thenCompose `method to achieve this.

Comment (hernaldourbina) [reply engine/src/main/java/com/odysseusinc/arachne/executionengine/execution/r/DockerOverseer.java:77]: Thanks. For this purpose, we will not be using generic exception then and will use specific.

**Comment by dmitrys-odysseus** [comment]
Defect: Still doesn't work for the case when container is being pulled.
The expected behaviour is not to start the container.
What currently happens is that container starts and then stopping it is attempted.

File: engine/src/main/java/com/odysseusinc/arachne/executionengine/execution/r/DockerOverseer.java [hunk 12]
```diff
@@ -58,17 +59,36 @@ public DockerOverseer(
...
         this.init = init;
     }
 
     @Override
     public CompletableFuture<ExecutionOutcome> abort() {
-        return result.isDone() ? result : CompletableFuture.completedFuture(
-                new ExecutionOutcome(Stage.ABORT, "Abort is not yet supported for docker analysis", null)
-        );
+        if (!init.isDone()) {
+            return CompletableFuture.completedFuture(new ExecutionOutcome(Stage.ABORTED, null, "Container initialization was canceled before it started."));
```
Comment (dmitrys-odysseus) [inline engine/src/main/java/com/odysseusinc/arachne/executionengine/execution/r/DockerOverseer.java:71]: This method still does nothing when init is not done. It does not prevent container from being started, e.g. the original defect https://github.com/OHDSI/ArachneExecutionEngine/pull/371#issuecomment-1894968994 still stands

**Review by dmitrys-odysseus — CHANGES_REQUESTED** [review]

File: engine/src/main/java/com/odysseusinc/arachne/executionengine/execution/r/DockerOverseer.java (hunk 12 above)
Comment (hernaldourbina) [reply engine/src/main/java/com/odysseusinc/arachne/executionengine/execution/r/DockerOverseer.java:72]: The first check ensures that if `init `is not done, it immediately returns a completed `ExecutionOutcome `indicating that container initialization was canceled. This prevents the Docker container from being started if initialization hasn't completed. Then, we first check if the container is still running using the `isContainerRunning()` method. If the container is running, we attempt to stop it; otherwise, we return a message indicating that the container has already completed. This ensures that you don't attempt to stop a container that has already finished its execution.

Comment (dmitrys-odysseus) [reply engine/src/main/java/com/odysseusinc/arachne/executionengine/execution/r/DockerOverseer.java:72]: "This prevents the Docker container from being started if initialization hasn't completed." - I don't see how can this be true. In my understanding `init` spins up docker container. I see no code here that will interfere with that.

File: engine/src/main/java/com/odysseusinc/arachne/executionengine/execution/r/DockerOverseer.java [hunk 13]
```diff
@@ -58,17 +60,53 @@ public DockerOverseer(
...
+                return CompletableFuture.completedFuture(result.join());
+            } else {
+                return CompletableFuture.supplyAsync(() -> {
+                    try {
+                        boolean isContainerRunning = isContainerRunning(containerId);
+                        if (isContainerRunning) {
+                            client.stopContainerCmd(containerId).exec();
+                            return new ExecutionOutcome(Stage.ABORTED, null, "Docker container aborted successfully:\n" + stdout);
+                        } else {
+                            return new ExecutionOutcome(Stage.ABORTED, null, "Docker container has already completed.");
```
Comment (dmitrys-odysseus) [inline engine/src/main/java/com/odysseusinc/arachne/executionengine/execution/r/DockerOverseer.java:86]: stdout lost again

File: engine/src/main/java/com/odysseusinc/arachne/executionengine/execution/r/DockerOverseer.java [hunk 14]
```diff
@@ -58,17 +60,53 @@ public DockerOverseer(
...
+                        return new ExecutionOutcome(Stage.ABORT, "Error aborting Docker container: " + e.getMessage(), "Error aborting Docker container: " + e.getMessage() + "\n" + stdout);
+                    }
+                }, executor);
+            }
+        });
+    }
+
+    private boolean isContainerRunning(String containerId) {
+        try {
+            InspectContainerResponse containerInfo = client.inspectContainerCmd(containerId).exec();
```
Comment (dmitrys-odysseus) [inline engine/src/main/java/com/odysseusinc/arachne/executionengine/execution/r/DockerOverseer.java:102]: There is no point in running a separate command to check state. You still have to deal with stop errors.

File: engine/src/main/java/com/odysseusinc/arachne/executionengine/execution/r/DockerOverseer.java [hunk 15]
```diff
@@ -58,17 +60,53 @@ public DockerOverseer(
...
+                        boolean isContainerRunning = isContainerRunning(containerId);
+                        if (isContainerRunning) {
+                            client.stopContainerCmd(containerId).exec();
+                            return new ExecutionOutcome(Stage.ABORTED, null, "Docker container aborted successfully:\n" + stdout);
+                        } else {
+                            return new ExecutionOutcome(Stage.ABORTED, null, "Docker container has already completed.");
+                        }
+                    } catch (NotFoundException e) {
+                        log.error("Container not found: {}", e.getMessage());
+                        return new ExecutionOutcome(Stage.ABORT, "Container not found: " + e.getMessage(), "Container not found: " + e.getMessage() + "\n" + stdout);
```
Comment (dmitrys-odysseus) [inline engine/src/main/java/com/odysseusinc/arachne/executionengine/execution/r/DockerOverseer.java:90]: ABORT is not a correct state here. Container completed on its own. Just join() on result instead...

File: engine/src/main/java/com/odysseusinc/arachne/executionengine/execution/r/DockerOverseer.java (hunk 12 above)
Comment (hernaldourbina) [reply engine/src/main/java/com/odysseusinc/arachne/executionengine/execution/r/DockerOverseer.java]: As our conversation on Friday, we are now using `CompletableFuture.cancel()` to stop the underlying process if it's already started, this will attempt to cancel the future but also, we are ensuring that a running process is stopped.

File: engine/src/main/java/com/odysseusinc/arachne/executionengine/execution/r/DockerOverseer.java [hunk 16]
```diff
@@ -54,21 +57,48 @@ public DockerOverseer(
...
-        );
+        if (!init.isDone()) {
+            init.cancel(true);
+        }
+
+        return init.handle((containerId, throwable) -> {
+            if (containerId == null) {
+                if (throwable != null) {
+                    log.error("Error during initialization: {}", throwable.getMessage());
+                    return new ExecutionOutcome(Stage.ABORTED, "Initialization failed: " + throwable.getMessage() + "\n" + stdout, "Initialization failed: " + throwable.getMessage() + "\n" + stdout);
```
Comment (dmitrys-odysseus) [inline engine/src/main/java/com/odysseusinc/arachne/executionengine/execution/r/DockerOverseer.java:80]: If initialization failed, the outcome stage should be INITIALIZE, not ABORTED

File: engine/src/main/java/com/odysseusinc/arachne/executionengine/execution/r/DockerOverseer.java [hunk 17]
```diff
@@ -54,21 +57,48 @@ public DockerOverseer(
...
+            } else {
+                if (result.isDone()) {
+                    return result.join();
+                } else {
+                    try {
+                        client.stopContainerCmd(containerId).exec();
+                        return new ExecutionOutcome(Stage.ABORTED, null, "Docker container aborted successfully:\n" + stdout);
+                    } catch (NotFoundException e) {
+                        log.info("Container not found or already stopped: " + e.getMessage());
+                        return new ExecutionOutcome(Stage.COMPLETED, "Container not found or already stopped: " + e.getMessage(), "Container not found or already stopped: " + e.getMessage() + "\n" + stdout);
```
Comment (dmitrys-odysseus) [inline engine/src/main/java/com/odysseusinc/arachne/executionengine/execution/r/DockerOverseer.java:93]: Always reporting COMPLETED + error is incorrect here. 
Could have completed successfully or failed. Use result.join() instead.

File: engine/src/main/java/com/odysseusinc/arachne/executionengine/execution/r/DockerOverseer.java [hunk 18]
```diff
@@ -54,21 +57,48 @@ public DockerOverseer(
...
                         : new ExecutionOutcome(Stage.EXECUTE, "Exit code " + exitCode, out);
-            }
+            } 
         }));
         pos = stdout.length();
         this.client = client;
-        init.thenAccept(containerId ->
-                executor.scheduleWithFixedDelay(this::writeLogs, updateInterval, updateInterval, TimeUnit.MILLISECONDS)
-        );
+        init.thenAccept(containerId -> {
```
Comment (dmitrys-odysseus) [inline engine/src/main/java/com/odysseusinc/arachne/executionengine/execution/r/DockerOverseer.java:64]: Please revert unnecessary formatting changes in constructor.

**Review by dmitrys-odysseus — CHANGES_REQUESTED** [review]
State reporting is still incorrect

**Review by dmitrys-odysseus — APPROVED** [review]
